### DISCUSSION TITLE
Fake/Mock an async event or isr?

### DISCUSSION THREAD

**Discussion by jnz86** [ThrowTheSwitch/Ceedling#708] [Q&A]
I'd like to fake an ISR that my UART is transmit is complete. I have a interrupt driven UART that sets a tx_busy flag cleared in the ISR on complete. This all works fine, but I'm having trouble simulating this for testing and coverage.

What I think I want is a function I can call saying in n milliseconds, call this callback (my ISR). This wouldn't be in any way real, but would let me get into my blocking section and then in some time have the callback fire and release the blocking. 

Is there a clean way to get this done without some large co-routine dependency?

**Comment by Letme** [comment]
Clean way is to have condition to while loop mockable, so that you can either inject custom fuction calls when false, and/or break the loop to test execution after the loop.
Your idea adds time to test case execution which usually amounts to alot in bigger programs, so i would rather have loop run 3 times, before i call isr, and then break out of the loop.
What you are explaining here is also a bit inapropriate unit. You can test and call isr outside of the blocking loop (its just a function) and test all options like that, then test the blocking loop separately.

The tools for async events usually need to be dynamic (runtime) but you have few static analysis like Polyspace Bugfinder or Synopsys Coverity, which check also locking of your shared variables and memory space (one bombards code with isr, other does statistical checkup- you locked once, but not other times).

**Reply by jnz86** [reply]
>Your idea adds time to test case execution which usually amounts to a lot in bigger programs, so i would rather have loop run 3 times, before i call isr, and then break out of the loop.

Yea, not exactly what I wanted to test. But I can shift my blocking function from a while(condition) to return busy and loop on that later.

Makes sense, was making sure I wasn't missing something easy.
